ML.Net and Tensorflow integration demo.

zeahmed opened the pull request:
In this demo, I created a `TensorflowTransform` that takes following as input and  creates column based on output name provided. The output column can be used as features for further training in the ML.Net pipeline.

- Model Directory
- Output Name (which become column in ML.Net)
- Input Name (it should be same as name in the model and name of the column in ML.Net)

zeahmed committed:
ML.Net and Tensorflow integration demo.

zeahmed committed:
Merge remote-tracking branch 'upstream/master' into tensorflow

ericstj commented on src/Microsoft.ML.Transforms/Microsoft.ML.Transforms.csproj:
```diff
@@ -6,6 +6,10 @@
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU'">
+    <NoWarn>CS8002;1591</NoWarn>
```
Interesting that it isn't strong named.  That will cause issues on desktop.  We could ask for it to be strong named, or build our own copy.

Also this can be
```
<NoWarn>$(NoWarn);CS8002</NoWarn>
```

And placed in the main PropertyGroup @ line 7. #Resolved

ericstj replied:
FYI this is fixed with my changes [https://github.com/zeahmed/machinelearning/pull/4](https://github.com/zeahmed/machinelearning/pull/4)

---
In reply to: [209299533](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209299533) [](ancestors = 209299533)

Zruty0 commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,410 @@
+        /// <summary>
+        /// Tensorflow session object
+        /// </summary>
+        private readonly TFSession _session;
```
>_session [](start = 35, length = 8)

Is TFSession thread-safe? #Closed

zeahmed replied:
TFSeesion and Graph seems to be threadsafe. I will look into Run method in python tensorflow code to see how they are ensuring thread safety.

---
In reply to: [209305362](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209305362) [](ancestors = 209305362)

Zruty0 commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,410 @@
+
+            return (ref VBuffer<float> dst) =>
+                {
+                    var runner = _session.GetRunner();
```
>runner [](start = 24, length = 6)

Can you have multiple runners at the same time from the same session?

It feels to me that the runner should be external to the getter delegate, and reused between calls. If this can not be done, explain why. #Resolved

zeahmed replied:
It seems like we can not have single runner mostly because of implementation of `Runner` which currently does not allow to set the input value once it has been set (using `AddInput` method). This is the limitation of `TensorflowSharp`.

---
In reply to: [209305767](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209305767) [](ancestors = 209305767)

Zruty0 replied:
we will still need to think of what functionality can be extracted out of the delegate.
For example, `input.Schema.GetColumnName(info.SrcIndices[i])` is not going to change from call to call

---
In reply to: [209413407](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209413407) [](ancestors = 209413407,209305767)

ericstj replied:
You can and should have multiple Runners, the Runner isn't really a heavy object, it's just a holder for the state of a run.  As such it doesn't make much sense to try to share it between runs.

---
In reply to: [209663760](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209663760) [](ancestors = 209663760,209413407,209305767)

Zruty0 commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,410 @@
+
+        public override void Save(ModelSaveContext ctx)
+        {
+            throw new NotImplementedException();
```
>throw [](start = 12, length = 5)

Have you given some thought on how we are planning to save/load such transforms?  #Closed

zeahmed replied:
I will look into it. I haven't thought about it yet. Maybe I will look into CNTK implementation in TLC to see what they are doing.

---
In reply to: [209306191](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209306191) [](ancestors = 209306191)

Zruty0 commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,410 @@
+            var status = new TFStatus();
+
+            return new TFSession().FromSavedModel(
+                    sessionOptions,
```
Is it safe to remove the TF model files after the session has read them? Or does the session defer reading till some later time? #Closed

zeahmed replied:
No, session loads the file immediately upon calling this method.

---
In reply to: [209306586](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209306586) [](ancestors = 209306586)

ericstj commented on test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:
```diff
@@ -0,0 +1,113 @@
+                                         b = new[] { 3.0f, 3.0f,
+                                                     3.0f, 3.0f } } }));
+
+                var trans = new TensorflowTransform(env, loader, model_location, "c", "a", "b");
```
I would hope that we don't need to explicitly declare this but can instead read it from model.  We should try to auto-map the IDV column names to TF input names.  Only when we can't map them should we require some user input. #Pending

Zruty0 replied:
But this will completely preclude any compile-time validation.

---
In reply to: [209317833](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209317833) [](ancestors = 209317833)

ericstj replied:
I don't see how listing strings without type information provides compile time validation. Sure if the pigsty tuples are used then I'd expect those to show up and just flow through the pipeline and work assuming remappings don't need to happen. I'd only expect a caller to explicitly list column names here if they need to change them from tuples that have already been defined in the data.

---
In reply to: [209324258](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209324258) [](ancestors = 209324258,209317833)

Zruty0 replied:
What I meant is, if you make transform work with no parameters whatsoever, and ONLY transform the columns it finds in the model, then Pigsty has nothing to 'attach to'.

---
In reply to: [209329418](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209329418) [](ancestors = 209329418,209324258,209317833)

Zruty0 replied:
But I think you're right, one doesn't preclude the other.

---
In reply to: [209374964](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209374964) [](ancestors = 209374964,209329418,209324258,209317833)

zeahmed replied:
@Eric, it could be but its not straight forward. Because there could be many variables in the tensorflow model and its difficult to identify which ones are inputs and need values to be fed. 

I will investigate it further if its possible. Because when saving model in tensorflow for serving, it ask for input and ouput. Let see if TFSharp has any convience to get it.

---
In reply to: [209375203](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209375203) [](ancestors = 209375203,209374964,209329418,209324258,209317833)

ericstj replied:
Sonoma does this FYI, you can look at how they did it.

---
In reply to: [209413718](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209413718) [](ancestors = 209413718,209375203,209374964,209329418,209324258,209317833)

zeahmed committed:
Got output in flatten array. Shapes and types are preloaded.

zeahmed commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,441 @@
+            var result = new float[size];
+            Marshal.Copy(data, result, 0, size);
+            return result;
+        }
```
Copying output buffer from `TFTensor` to managed memory. #Resolved

ericstj replied:
This is OK as a temporary hack, put in a big comment here (or some asserts) about how this only works for row-major float tensors.

---
In reply to: [209413803](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209413803) [](ancestors = 209413803)

Zruty0 commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,441 @@
+using TensorFlow;
+
+[assembly: LoadableClass(TensorflowTransform.Summary, typeof(TensorflowTransform), typeof(TensorflowTransform.Arguments), typeof(TensorflowTransform),
+    TensorflowTransform.UserName, "CopyColumns", "CopyColumnsTransform", TensorflowTransform.ShortName, DocName = "transform/CopyColumnsTransform.md")]
```
>CopyColumns [](start = 35, length = 11)

by the way, you need to fix the attribute #Resolved

Zruty0 commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,441 @@
+
+namespace Microsoft.ML.Transforms
+{
+    public sealed class TensorflowTransform : RowToRowMapperTransformBase, IDisposable
```
>IDisposable [](start = 75, length = 11)

That is incorrect. Our transforms should never be disposable. #Resolved

ericstj replied:
Took a look at DnnImageFeaturizerTransform and it just lets GC finalize disposable CNTK.Function. Also checked CntkPredictorBase and it does the same for its public Function Model property.

---
In reply to: [209664938](https://github.com/zeahmed/machinelearning/pull/3#discussion_r209664938) [](ancestors = 209664938)

Zruty0 commented:
@yaeld is added to the review. #Closed

Zruty0 commented:
@yaeldekel is added to the review. #Closed

ericstj committed:
Don't use TensorFlowSharp, instead carry our own subset of the TF# API

This is just a stub to begin iterating on.

To follow:
1. Pull over a copy of TF# source
2. Tree-shake unused code.
3. Rewrite / Rename TF# API as appropriate.
4. Come up with a strategy for providing TF bins (using TF#'s copy now, merely as a starting point).

ericstj committed:
Bring over copy of TF# source

*NOT BUILDING*

ericstj committed:
Add license to TF# files and ifdef to make them build

zeahmed committed:
Added model serialization to disk using Tensorflow frozen model scheme.

zeahmed commented on test/Microsoft.ML.Tests/ScenariosWithDirectInstantiation/TensorflowTests.cs:
```diff
@@ -0,0 +1,173 @@
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 139, 253, 190, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 11, 190, 253, 70, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 35, 241, 225, 160, 108, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 81, 240, 253, 253, 119, 25, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 45, 186, 253, 253, 150, 27, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 93, 252, 253, 187, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  ...
+                };
```
Serialized the TF model in ML.NET which enables us to use the pipeline for scoring and prediction now...:blush:  #Resolved

zeahmed commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,501 @@
+                {
+                    dims[k] = (int)(shape[k] == -1 ? batchSize : shape[k]);
+                }
+                // Output tensor is expected to be of shape [Batch, Data]
```
>// Output tensor is expected to be of shape [Batch, Data] [](start = 16, length = 57)

Not any more. I will remove the comment. #Resolved

zeahmed commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,501 @@
+        /// <summary>
+        /// Tensorflow session object
+        /// </summary>
+        private TFSession _session;
```
>_session [](start = 26, length = 8)

Need to find out a way to dispose this safely. #Resolved

yaeldekel replied:
Instead of implementing IDisposable, you can dispose of this resource in the finalizer of the transform.

---
In reply to: [210042898](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210042898) [](ancestors = 210042898)

zeahmed commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,501 @@
+            ctx.SaveBinaryStream("TFModel", w =>
+            {
+                w.WriteByteArray(buffer.ToArray());
+            });
```
check for success here. #Resolved

yaeldekel replied:
ToGraphDef might throw exceptions, but otherwise, could this not succeed?

---
In reply to: [210043097](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210043097) [](ancestors = 210043097)

zeahmed replied:
To me comment shows at different line number after the several commits. Originally, I commented while loading at following code.

```
byte[] data = null;
ctx.TryLoadBinaryStream("TFModel", r =>
{
       data = r.ReadByteArray();
});
```

---
In reply to: [210651836](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210651836) [](ancestors = 210651836,210043097)

ericstj committed:
Trim down TensorFlowSharp API

ericstj committed:
Enable StyleCop and fix issues

ericstj committed:
Enable ML Codeanalyzer

I've fixed some minor rule issues and disabled the naming analyzers since those churn the source too much.

ericstj committed:
Move TensorFlow support into Microsoft.ML.Transforms

For now it is BYOTF (bring your own TensorFlow) so the tests are grabbing the TF bits for TF#.

ericstj committed:
Put TensorFlow support under Microsoft.ML.Transforms.TensorFlow and make internal

zeahmed committed:
Added support for different types.

zeahmed commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,681 @@
+
+            if (tfType == TFDataType.Int32)
+            {
+                return CreateTensorValueGetter<DvInt4>(input, type, colIndex, tfShape);
```
>DvInt4 [](start = 47, length = 6)

ML.Net types cannot be passed to TF right now. #Pending

zeahmed replied:
There is work going on on removing ML.Net types and using pure .net types. That will resolve this issue.

---
In reply to: [210172882](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210172882) [](ancestors = 210172882)

yaeldekel replied:
As long as this work is not done yet, we should either not support these types, or use the ML.NET types and convert to .net types before passing to TF. this method gets the getters of the inputs, and will throw exception if called with Int64, bool or string.

---
In reply to: [210456041](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210456041) [](ancestors = 210456041,210172882)

zeahmed commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,681 @@
+            {
+                var scalar = default(T);
+                _srcgetter(ref scalar);
+                return (TFTensor)_ctor.Invoke(null, new object[] { scalar });
```
>(TFTensor)_ctor.Invoke(null, new object[] { scalar }); [](start = 23, length = 54)

This reflection call is not intended here. We need a method in TFSharp where we can pass in generic data to TF. Currently, there are methods in TFSharp to do that but those are not exposed. I and @Eric will work on that. #Resolved

zeahmed commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,681 @@
+                _srcgetter(ref tmpBuf);
+                var dense = default(VBuffer<T>);
+                tmpBuf.CopyToDense(ref dense);
+                return (TFTensor)_method.Invoke(null, new object[] { _tfShape, dense.Values, 0, dense.Length });
```
>return (TFTensor)_method.Invoke(null, new object[] { _tfShape, dense.Values, 0, dense.Length }); [](start = 16, length = 96)

This reflection call is not intended here. We need a method in TFSharp where we can pass in generic data to TF. Currently, there are methods in TFSharp to do that but those are not exposed. I and @Eric will work on that. #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,681 @@
+            }
+            public TFTensor GetTensor()
+            {
+                var tmpBuf = default(VBuffer<T>);
```
>tmpBuf [](start = 20, length = 6)

This class should have a VBuffer as a private field, otherwise a new array will be allocated on every call. #Resolved

yaeldekel replied:
Alternatively, you can make ITensorValueGetter generic on "T", and the GetTensor method will be GetTensor(ref T input).

---
In reply to: [210415541](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210415541) [](ancestors = 210415541)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,681 @@
+            {
+                Func<IRow, int, ValueGetter<VBuffer<float>>> srcDel = MakeGetter<float>;
+                var meth = srcDel.GetMethodInfo().GetGenericMethodDefinition().MakeGenericMethod(outInfo.ItemType.RawType);
+                return (Delegate)meth.Invoke(this, new object[] { input, iinfo });
```
You can make this call only once for all types.
Also, there is a utility method Utils.MarshalInvoke that does this, and provides compile time type checking. #Resolved

ericstj committed:
Create a generic method for creating TF tensors

This method will accept a .NET type and produce a tensor.

In the case of a scalar it allocates a single element on the managed heap to back the tensor.

In the case of single or multi-dimensional array it pins the array on the managed heap and wraps it in a tensor, allowing TF to directly interact with the managed array.

In all other cases we throw.

To produce this method I used T4 templates, which autogenerate the CS source when you edit the tt file.  See https://docs.microsoft.com/en-us/visualstudio/modeling/design-time-code-generation-by-using-t4-text-templates#creating-a-design-time-t4-text-template for more details.  We use the same approach in CoreFx and other repos for dealing with type-specific functionality breakout from a generic method.

zeahmed committed:
Merge pull request #4 from ericstj/wrapTF

[WIP] Create a minimal TF binding

zeahmed committed:
Addressed reviewers' comments.

ericstj commented on src/Microsoft.ML.Transforms/TensorFlow/TensorGeneric.cs:
```diff
@@ -0,0 +1,140 @@
+        /// <typeparam name="T">.NET type of tensor to create</typeparam>
+        /// <param name="data">value of tensor</param>
+        /// <param name="shape">shape of tensor. Only valid when passing in arrays.</param>
+        public static TFTensor Create<T>(T data, TFShape shape = null)
```
>shape [](start = 57, length = 5)

Shape should be calculatable from Array, we shouldn't need to pass it down.  Are you doing this because you're trying to treat an array as a different shape tensor than its managed array's shape? #Resolved

zeahmed replied:
Yes, data is kept in flatten array as row major format while shape indicates the dimensions. I was thinking to make a different create for arrays. What do you think?

---
In reply to: [210640853](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210640853) [](ancestors = 210640853)

ericstj replied:
I see.  A different method makes sense.  Can you call it CreateReshape or something like that?  It can live parallel to Create and could take T[] and avoid the typeof(T).IsArray/GetElementType code.

As an aside: does it cause issue that ML.NET forces n-dimensional data down into 1-d row-major layout?  That constraint feels limiting and seems like it would lead to extra work when bringing together components (end-user apps, transforms, etc) that need n-d or different layouts.

---
In reply to: [210698485](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210698485) [](ancestors = 210698485,210640853)

zeahmed replied:
This transform will be some where in the middle of the pipeline. At low level, all the transform use flattened array with dimensions maintained in the IDV schema. There will be front facing transform such as image transform that will handle the multi dimension case (e.g. reading images from file in 3-d format) and format the data consumable by other transform.

---
In reply to: [210708110](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210708110) [](ancestors = 210708110,210698485,210640853)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+
+            private static string TestTypes(ColumnType[] types)
+            {
+                return null;
```
>null [](start = 23, length = 4)

The IDV type system has key types, DateTime, unsigned integer types, etc. I think this should make sure that the types are one of the types allowed above in Tf2MlNetType().

yaeldekel replied:
It also seems that the inputs must be vectors.

---
In reply to: [210646613](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210646613) [](ancestors = 210646613)

yaeldekel replied:
I am not sure I was right about this assumption - does TF accept scalar inputs?

---
In reply to: [210657676](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210657676) [](ancestors = 210657676,210646613)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+        ~TensorflowTransform()
+        {
+            _session.CloseSession();
+            _session.DeleteSession();
```
>_session [](start = 12, length = 8)

Should we also call _session.Dispose()? #Closed

zeahmed replied:
I see dispose already calls _session.DeleteSession();

---
In reply to: [210647126](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210647126) [](ancestors = 210647126)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+        /// <summary>
+        /// Tensorflow session object
+        /// </summary>
+        private TFSession _session;
```
>_session [](start = 26, length = 8)

This can be readonly. #Closed

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+            for (int i = 0; i < args.Column.Length; i++)
+                Host.CheckUserArg(Utils.Size(args.Column[i].Source) > 0, nameof(args.Column));
+
+            _batchSize = args.BatchSize;
```
>BatchSize [](start = 30, length = 9)

Check that this is positive. #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+        {
+            Host.AssertValue(ctx);
+
+            _batchSize = ctx.Reader.ReadInt32();
```
>_batchSize [](start = 12, length = 10)

CheckDecode that this is positive. In the save method assert that it is positive. #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+                Host.CheckUserArg(Utils.Size(args.Column[i].Source) > 0, nameof(args.Column));
+
+            _batchSize = args.BatchSize;
+            _session = LoadTFSession(args.ModelFile);
```
>ModelFile [](start = 42, length = 9)

Check that the file exists. #Closed

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+        private TFSession LoadTFSession(string modelFile)
+        {
+            var graph = new TFGraph();
+            graph.Import(File.ReadAllBytes(modelFile), "");
```
>Import [](start = 18, length = 6)

This call can go in a try/catch, with the catch rethrowing the exception through the Host. You can add an argument to LoadTFSession to tell the Host whether to throw ExceptDecode() or ExceptUserArg().

yaeldekel replied:
I think this method can be used both in the ctor from arguments and the ctor from file, and take a bool parameter indicating what kind of exception to throw.

---
In reply to: [210650887](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210650887) [](ancestors = 210650887)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+                return new TensorValueGetter<T>(input, colIndex);
+        }
+
+        private ITensorValueGetter CreateTensorValueGetterVec(IRow input, TFDataType tfType, ColumnType type, int colIndex, TFShape tfShape)
```
>CreateTensorValueGetterVec [](start = 35, length = 26)

Add a method similar to Tf2MlNetType() that returns a Type, and then you can reduce this method to to lines:

var type = Tf2SystemType(tfType);
return Utils.MarshalInvoke(CreateTensorValueGetter<int>, type, input, type, colIndex, tfShape);

 #Resolved

zeahmed replied:
I changed what about the template type <int> in CreateTensorValueGetter<int>.

---
In reply to: [210656910](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210656910) [](ancestors = 210656910)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+            public string ModelFile;
+
+            [Argument(ArgumentType.Required, HelpText = "Batch size.", ShortName = "BatchSize", SortOrder = 3)]
+            public int BatchSize = 1;
```
>BatchSize [](start = 23, length = 9)

What does BatchSize mean? Is it the number of examples scored in one call to TF? #Resolved

zeahmed replied:
Yes, If model has an unknown dimension then we assume that its a batch dimension and fill it with BatchSize. e.g. [?,2,2] so '?' is batch dimension.

---
In reply to: [210665223](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210665223) [](ancestors = 210665223)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+        {
+            Host.AssertValue(input);
+            // Need to check type here.
+            //Host.Assert(Infos[iinfo].TypeSrc.IsText);
```
>Assert [](start = 19, length = 6)

If you pass in a ColumnType as well here, then you can check that typeof(T)==columnType.ItemType.RawType. #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+
+            if (tfType == TFDataType.Int64)
+            {
+                return Utils.MarshalInvoke(MakeGetter<Int64>, outInfo.ItemType.RawType, input, iinfo);
```
>Int64 [](start = 54, length = 5)

Do you have any tests testing this (and int, bool and string)?  #Pending

zeahmed replied:
Not yet because ML.Net does not support dotnet types it uses DvInt etc. Once ZeeshanS changes are there I will be able to test.

---
In reply to: [210684078](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210684078) [](ancestors = 210684078)

yaeldekel replied:
I don't know when Zeeshan's changes will be done, but for now I don't think this would work. If you don't want to handle these types using DvInt,DvBool etc., then I think we should not support them until the type system changes go in.

---
In reply to: [210718189](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210718189) [](ancestors = 210718189,210684078)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+                    {
+                        var inputName = tfInfo.InputColNames[i];
+                        var type = info.SrcTypes[i];
+                        runner.AddInput(inputName, srcTensorGetters[i].GetTensor());
```
>GetTensor [](start = 71, length = 9)

Is there a way to make this method not allocate a new array on every call? #Resolved

yaeldekel replied:
Also, we need to keep track of the allocations and dispose of them by overriding the RowCursor's Dispose() method (IRowCursors are IDisposable).

---
In reply to: [210685450](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210685450) [](ancestors = 210685450)

zeahmed replied:
It wont allocate new array. It takes the handle of the .net array and pass it on to TF.

---
In reply to: [210688784](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210688784) [](ancestors = 210688784,210685450)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+                        runner.AddInput(inputName, srcTensorGetters[i].GetTensor());
+                    }
+
+                    var tensors = runner.Fetch(_bindings.OutputColNames[iinfo]).Run();
```
>Run [](start = 80, length = 3)

Same question for the outputs. #Resolved

zeahmed replied:
For output, its unmanage array that is returned and we make a copy of it after this line.

---
In reply to: [210685620](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210685620) [](ancestors = 210685620)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+
+                    Contracts.Assert(tensors.Length > 0);
+
+                    var output = FetchData<T>(tensors[0].Data, _bindings.OutputCols[iinfo].VectorSize);
```
>FetchData [](start = 33, length = 9)

FetchData should use the values array of dst instead of creating a new one (you can check to see the common pattern in other transforms: if the length of dst.Values is not sufficient, we create a new array.) #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+            return result;
+        }
+
+        private Delegate MakeGetter(IRow input, int iinfo)
```
>MakeGetter [](start = 25, length = 10)

Same comment as for CreateTensorValueGetterVec. #Resolved

zeahmed replied:
Done!

---
In reply to: [210687297](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210687297) [](ancestors = 210687297)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+                        runner.AddInput(inputName, srcTensorGetters[i].GetTensor());
+                    }
+
+                    var tensors = runner.Fetch(_bindings.OutputColNames[iinfo]).Run();
```
>tensors [](start = 24, length = 7)

Does TF know how to output values from multiple layers in one call? If this is the case, we should not make this transform ManyToOne, it should be many-to-many. #Pending

zeahmed replied:
Yes, we can get multiple outputs in one call.  I am not sure right now how to do it in ML.Net. This is the next thing I am going to focus on.

---
In reply to: [210687722](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210687722) [](ancestors = 210687722)

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+                {
+                    if (!activeInfos(iinfo))
+                        continue;
+                    getters[iinfo] = MakeGetter(input, iinfo);
```
>MakeGetter [](start = 37, length = 10)

Add an out parameter of a disposer to MakeGetter, this way it can be used here, and by the RowCursor. #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorFlow/Buffer.cs:
```diff
@@ -0,0 +1,211 @@
+
+namespace Microsoft.ML.Transforms.TensorFlow
+{
+	/// <summary>
```
>	 [](start = 0, length = 1)

Change the tabs to spaces. (in the other files from TFSharp as well). #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+
+            private static DataKind Tf2MlNetType(TFDataType type)
+            {
+                if(type == TFDataType.Float)
```
>if [](start = 16, length = 2)

Can this be changed to a switch statement? #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+
+            if (tfType == TFDataType.String)
+            {
+                return Utils.MarshalInvoke(MakeGetter<string>, outInfo.ItemType.RawType, input, iinfo);
```
>string [](start = 54, length = 6)

This should be DvText. #Resolved

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,665 @@
+
+            if (tfType == TFDataType.String)
+            {
+                return CreateTensorValueGetter<string>(input, type, colIndex, tfShape);
```
>string [](start = 47, length = 6)

Text will continue to be handled using DvText even after we move to .net types. #Resolved

zeahmed committed:
Addressed reviewers' comments.

ericstj commented on src/Microsoft.ML.Transforms/TensorFlow/TensorGeneric.cs:
```diff
@@ -0,0 +1,140 @@
+        /// <summary>
+        /// Creates a tensor representing type T[].
+        /// T[] will be pinned and wrapped in a tensor.
+        /// This method does not support jagged arrays since those require copies.
```
>This method does not support jagged arrays since those require copies. [](start = 12, length = 70)

This comment is no longer relevant since you only accept single dimensional arrays. #Resolved

ericstj commented on src/Microsoft.ML.Transforms/TensorFlow/TensorGeneric.cs:
```diff
@@ -0,0 +1,140 @@
+        /// </summary>
+        /// <typeparam name="T">.NET type of tensor to create</typeparam>
+        /// <param name="data">value of tensor</param>
+        public static TFTensor Create<T>(T data)
```
>Create [](start = 31, length = 6)

Consider changing this to createscalar. #Resolved

ericstj commented on src/Microsoft.ML.Transforms/TensorFlow/TensorflowUtils.cs:
```diff
@@ -0,0 +1,129 @@
+
+        internal static TFDataType MlNet2TfType(Type type)
+        {
+            if (type == typeof(float))
```
>( [](start = 30, length = 1)

This could be code-gen'ed by the TT file. #Resolved

zeahmed replied:
I was thinking that too. Let me see...:)

---
In reply to: [210747481](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210747481) [](ancestors = 210747481)

zeahmed replied:
Actually there is already a method  in TFSharp. I copied it.

---
In reply to: [210747970](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210747970) [](ancestors = 210747970,210747481)

ericstj replied:
We should still plan on deleting all this TF# code and code-gen'ing it.  It's all boiler-plate copy/paste that can be driven off the type mapping.

---
In reply to: [210758078](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210758078) [](ancestors = 210758078,210747970,210747481)

zeahmed replied:
Agreed! 

---
In reply to: [210760965](https://github.com/zeahmed/machinelearning/pull/3#discussion_r210760965) [](ancestors = 210760965,210758078,210747970,210747481)

zeahmed committed:
Removed batch size as configurable param and setting it to one for now.

zeahmed committed:
Added TensorTypeFromType from TFSharp.

yaeldekel commented on Microsoft.ML.sln:
```diff
@@ -103,8 +103,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.ML.CpuMath.UnitTe
 EndProject
-
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "Microsoft.ML.FSharp.Tests", "test\Microsoft.ML.FSharp.Tests\Microsoft.ML.FSharp.Tests.fsproj", "{802233D6-8CC0-46AD-9F23-FEE1E9AED9B3}"
+Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "Microsoft.ML.FSharp.Tests", "test\Microsoft.ML.FSharp.Tests\Microsoft.ML.FSharp.Tests.fsproj", "{802233D6-8CC0-46AD-9F23-FEE1E9AED9B3}"
```
>6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705 [](start = 10, length = 36)

Why did this change?

yaeldekel commented on src/Microsoft.ML.Transforms/TensorFlow/Tensor.cs:
```diff
@@ -0,0 +1,952 @@
+namespace Microsoft.ML.Transforms.TensorFlow
+{
+
+	/// <summary>
```
>	 [](start = 0, length = 1)

Convert tabs to spaces.

yaeldekel commented on src/Microsoft.ML.Transforms/TensorFlow/Tensorflow.cs:
```diff
@@ -0,0 +1,2147 @@
+
+namespace Microsoft.ML.Transforms.TensorFlow
+{
+	internal static partial class NativeBinding
```
>	 [](start = 0, length = 1)

Convert tabs to spaces

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+            _bindings.Save(ctx);
+        }
+
+        public void DisposeTFSession()
```
>DisposeTFSession [](start = 20, length = 16)

I think this method can be removed.

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+{
+    public sealed class TensorflowTransform : RowToRowMapperTransformBase
+    {
+        public sealed class Column : ManyToOneColumn
```
>ManyToOneColumn [](start = 37, length = 15)

Consider not using this base class (or the ManyToOneColumnBindingsBase class). These are intended for applying the transform simultaneously to multiple sets of input columns. It is advantageous in case the transform requires training (such as ngram transform) because defining multiple transforms would require multiple passes over the data.

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+            }
+        }
+
+        public sealed class Arguments : TransformInputBase
```
>Arguments [](start = 28, length = 9)

I would change the structure of these arguments to be an array of strings for the inputs, and a string for the output.

Later we may want to change the arguments to have a mapping from the IDV column names to the model input names, and/or from the model output names to IDV output column names.

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+            public readonly ColumnType[] OutputCols;
+            public readonly TFDataType[] OutputTFTypes;
+
+            public Bindings(Column[] columns, ISchema schemaInput, TensorflowTransform parent)
```
>Bindings [](start = 19, length = 8)

Change this class to have just one private ctor, and two public ctors/create methods that call the private one. This way if we make any changes to this class we won't need to remember to make them in two places.

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+            _session.DeleteSession();
+        }
+
+        private ValueGetter<T> GetSrcGetter<T>(IRow input, int iinfo, int isrc)
```
>GetSrcGetter [](start = 31, length = 12)

Is this used anywhere?

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+        private ITensorValueGetter CreateTensorValueGetterVec(IRow input, TFDataType tfType, ColumnType columnType, int colIndex, TFShape tfShape)
+        {
+            var type = TFTensor.TypeFromTensorType(tfType);
+            if (type != null)
```
>type [](start = 16, length = 4)

We should be able to assert here that type != null.

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+                return new TensorValueGetter<T>(input, colIndex);
+        }
+
+        private ITensorValueGetter CreateTensorValueGetterVec(IRow input, TFDataType tfType, ColumnType columnType, int colIndex, TFShape tfShape)
```
>Vec [](start = 58, length = 3)

Can the "Vec" suffix be removed? I don't see a method that is for "NotVec".

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+                    Contracts.Assert(tensors.Length > 0);
+
+                    var values = dst.Values;
+                    if (Utils.Size(values) != _bindings.OutputCols[iinfo].VectorSize)
```
>!= [](start = 43, length = 2)

Doesn't need to be exactly equal. < is enough here.

yaeldekel commented on src/Microsoft.ML.Transforms/TensorflowTransform.cs:
```diff
@@ -0,0 +1,601 @@
+                        values = new T[_bindings.OutputCols[iinfo].VectorSize];
+
+                    TensorflowUtils.FetchData<T>(tensors[0].Data, values);
+                    dst = new VBuffer<T>(values.Length, values);
```
>VBuffer [](start = 30, length = 7)

Pass dst.Indices to the new VBuffer as well.
